### ISSUE TITLE
@Only annotation

### ISSUE THREAD

**Issue by cezarykluczynski** [testng-team/testng#818]
Feature request. @Only annotation would be similiar to Mocha's only method: https://mochajs.org/#exclusive-tests

Whenever a @Only annotation is found anywhere in a project, only those classes and methods that have it, will be executed.

What do you think?

**Comment by cbeust** [comment]
Interesting, I like the idea. I'm guessing the goal is for when you work on individual test methods?

Right now, I create a `testng-single.xml` that contains just that test but your suggestion is much more convenient, thanks!

**Comment by rschmitt** [comment]
IntelliJ already supports this (Eclipse probably does as well), and I'm not sure where you'd need a non-interactive version of this functionality.

**Comment by cezarykluczynski** [comment]
@cbeust Yes, that's the goal.

@rschmitt Not everyone uses IDE (right now, I mix IntelliJ with Sublime), and not every environment is interactive, like a CI server. I guess that's just a different way of doing things that can done using a separate XML config or a "maven test -Dtest=ClassName" command.

**Comment by verhagen** [comment]
@cezarykluczynski this @Only would be kind of counterpart of @Ignore ?

**Comment by cezarykluczynski** [comment]
@verhagen That's right, instead of removing an annoted test from the list of tests to execute, like @Ignore does, it would construct a list of tests to execute from only those tests that are annotated with @Only.

**Comment by juherr** [comment]
It would be interesting to be able to configure the behavior about `dependsOnX` dependencies: it seems that some persons want sometimes to not follow them.
See: https://youtrack.jetbrains.com/issue/IDEA-141621

**Comment by cbeust** [comment]
That's a strange scenario: if your test can run fine without its dependencies, why does it have any in the first place?

For what it's worth, `testng-failed.xml` solves this problem by calculating the transitive closure of the dependencies and including them in the xml file. And as far as I remember, the Eclipse plug-in uses that same logic to do the same thing (so you can run individual methods, even if they have dependencies).

This seems like a more useful approach than just removing these dependencies.

**Comment by juherr** [comment]
@cbeust You should say it on https://youtrack.jetbrains.com/issue/IDEA-141073 . Maybe, it will help to fix #881 quicker

**Comment by cbeust** [comment]
Done.

**Comment by juherr** [comment]
:+1:

**Comment by juherr** [comment]
Triage verdict: **still relevant**.

Current TestNG still has no `@Only`/focused-test annotation or equivalent framework-level exclusive-selection marker. The existing mechanisms (XML method selection, groups, runner/IDE selection and method selectors) require the user or runner to specify what to run; they do not implement the requested semantics where the mere presence of an annotation anywhere changes the run-wide selection set. I found no `GITHUB-818`, merged/open PR, or newer canonical issue implementing it.

This remains a design enhancement because semantics need to be decided for dependencies, class-level vs method-level markers, factories/data providers, multiple suites, and whether CI should be able to reject accidentally committed focused tests.
